src/cunumeric/index/advanced_indexing.cu

```diff
@@ -29,95 +29,113 @@ template <typename Output, typename Pitches, typename Point, int32_t DIM>
+                       const size_t key_dim)
 {
   size_t value = 0;
-  for (size_t idx = 0; idx < iters; idx++) {
-    const size_t offset = (idx * gridDim.x + blockIdx.x) * blockDim.x + threadIdx.x;
-    if (offset < volume) {
-      auto point      = pitches.unflatten(offset, origin);
-      auto val        = static_cast<size_t>(index[point]);
-      offsets[offset] = val;
-      SumReduction<size_t>::fold<true>(value, val);
+  for (size_t i = 0; i < iters; i++) {
+    size_t idx = (i * gridDim.x + blockIdx.x) * blockDim.x + threadIdx.x;
+    if (idx > volume) break;
+    auto point = pitches.unflatten(idx, origin);
+    size_t val;
+    if (index[point] and ((idx + 1) % skip_size == 0)) {
+      offsets[idx] = 1;
+      val          = 1;
+    } else {
+      offsets[idx] = 0;
+      val          = 0;
     }
+    SumReduction<size_t>::fold<true>(value, val);
   }
   // Every thread in the thread block must participate in the exchange to get correct results
   reduce_output(out, value);
 }
 
-template <typename VAL, int DIM1, int DIM2>
+template <typename VAL, int DIM>
 static __global__ void __launch_bounds__(THREADS_PER_BLOCK, MIN_CTAS_PER_SM)
   advanced_indexing_kernel(size_t volume,
-                           AccessorRO<VAL, DIM1> in,
-                           AccessorRO<bool, DIM2> index,
-                           Buffer<VAL> out,
-                           Pitches<DIM1 - 1> pitches_input,
-                           Point<DIM1> origin_input,
-                           Pitches<DIM2 - 1> pitches_index,
-                           Point<DIM2> origin_index,
-                           Buffer<int64_t> offsets)
+                           AccessorRO<VAL, DIM> in,
+                           AccessorRO<bool, DIM> index,
+                           Buffer<VAL, DIM> out,
+                           Pitches<DIM - 1> pitches,
+                           Point<DIM> origin,
+                           Buffer<int64_t> offsets,
+                           const size_t skip_size,
+                           const size_t key_dim)
 {
-  // FIXME works only when DIM1==DIM2
   const size_t tid = blockIdx.x * blockDim.x + threadIdx.x;
-  if (tid >= volume) return;
-  auto point       = pitches_index.unflatten(tid, origin_index);
-  auto point_input = pitches_input.unflatten(tid, origin_input);
+  if (tid > volume) return;
+  auto point = pitches.unflatten(tid, origin);
   if (index[point] == true) {
-    int64_t offset = offsets[tid];
-    out[offset]    = in[point_input];
+    Point<DIM> out_p;
+    out_p[0] = offsets[tid];
+    for (size_t i = 0; i < DIM - key_dim; i++) {
+      size_t j     = key_dim + i;
+      out_p[i + 1] = point[j];
+    }
+    for (size_t i = DIM - key_dim + 1; i < DIM; i++) out_p[i] = 0;
+    out[out_p] = in[point];
   }
 }
 
-template <typename VAL, int DIM1, int DIM2>
+template <typename VAL, int DIM>
 static __global__ void __launch_bounds__(THREADS_PER_BLOCK, MIN_CTAS_PER_SM)
   advanced_indexing_kernel(size_t volume,
```

Comment:
like I said previously, this kernel can be deduplicated using some template foo.